core/trace.c

```diff
@@ -561,4 +541,140 @@ void stop_trace() {
+        NULL,   // void* pointer,
+        tag,    // tag* tag,
+        -1,     // int worker, // no worker ID needed because this is called within a mutex
+        fed_id, // int src_id,
+        -1,     // int dst_id,
+        NULL,   // instant_t* physical_time (will be generated)
+        NULL,   // trigger_t* trigger,
+        0       // interval_t extra_delay
+    );
+}
+
+/**
+ * Trace federate receiving a message from the RTI.
+ * @param event_type The type of event. Possible values are:
+ *
+ * @param fed_id The federate identifier.
+ * @param tag Pointer to the tag that has been received, or NULL.
+ */
+void tracepoint_federate_from_RTI(trace_event_t event_type, int fed_id, tag_t* tag) {
+    // trace_event_t event_type = (type == MSG_TYPE_TAG_ADVANCE_GRANT)? federate_TAG : federate_PTAG;
+    tracepoint(event_type,
+        NULL,   // void* pointer,
+        tag,    // tag* tag,
+        -1,     // int worker, // no worker ID needed because this is called within a mutex
+        fed_id, // int src_id,
+        -1,     // int dst_id,
+        NULL,   // instant_t* physical_time (will be generated)
+        NULL,   // trigger_t* trigger,
+        0       // interval_t extra_delay
+    );
+}
+
+/**
+ * Trace federate sending a message to another federate.
+ * @param event_type The type of event. Possible values are:
+ *
+ * @param fed_id The federate identifier.
+ * @param partner_id The partner federate identifier.
+ * @param tag Pointer to the tag that has been sent, or NULL.
+ */
+void tracepoint_federate_to_federate(trace_event_t event_type, int fed_id, int partner_id, tag_t *tag) {
+    tracepoint(event_type,
+        NULL,   // void* pointer,
+        tag,    // tag* tag,
+        -1,     // int worker, // no worker ID needed because this is called within a mutex
+        fed_id, // int src_id,
+        partner_id,     // int dst_id,
+        NULL,   // instant_t* physical_time (will be generated)
+        NULL,   // trigger_t* trigger,
+        0       // interval_t extra_delay
+    );
+}
+
+/**
+ * Trace federate receiving a message from another federate.
+ * @param event_type The type of event. Possible values are:
+ *
+ * @param fed_id The federate identifier.
+ * @param partner_id The partner federate identifier.
+ * @param tag Pointer to the tag that has been received, or NULL.
+ */
+void tracepoint_federate_from_federate(trace_event_t event_type, int fed_id, int partner_id, tag_t *tag) {
+    tracepoint(event_type,
+        NULL,   // void* pointer,
+        tag,    // tag* tag,
+        -1,     // int worker, // no worker ID needed because this is called within a mutex
+        fed_id, // int src_id,
+        partner_id,     // int dst_id,
+        NULL,   // instant_t* physical_time (will be generated)
+        NULL,   // trigger_t* trigger,
+        0       // interval_t extra_delay
+    );
+}
+#endif // FEDERATED
+
+////////////////////////////////////////////////////////////
+//// For RTI execution
```

Comment:
This will make it harder to eventually disentangle the RTI from `reactor-c` (IIRC, a long-standing to-do is to put the RTI in a separate repository). Can these function implementations be factored out into a file in `core/federated/RTI` that is only used in the building of the RTI?

The same comment applies to the RTI-specific changes to `trace.h`. If we disentangle these things then that might allow us to avoid having to use both `LF_TRACE` and `RTI_TRACE` since the build system would take care of linking in the right code for the given executable.

Reply:
I agree with Peter. Let's handle this now, before merging.